[TBN-1136] Added a skill for async-api (#1264)
wisemen-digital/vue-core: packages/api/nestjs-async-api/skills/async-api/SKILL.md

---
name: async-api
description: Create documentation for cloud events. Use this when adding a (NATS) integration event.
---

Generate AsyncAPI 3.0.0 specification documents from type-safe channel definitions and Swagger-decorated DTOs.

# Validate event structure
- Make sure the properties of the event are decorated using '@nestjs/swagger'. 

# Define a channel
- Define the channel at the bottom of the same file as the integration event definition.
- Create a constant for the subject (i.e. topic) where the event will be sent
  - The subject may contain parameters like 'environment', 'uuid', ... These parameters must be enclosed by curly brackets (`{}`)
- Create a channel using the `createChannel` method from the '@wisemen/nestjs-async-api' package.
  - All parameters defined in the subject must be documented in the channel config param of the method.

>   - Define operations for the event. These can be 'send' or 'receive'

Reviewer: mss verduidelijken voor wat send en receive dient?

```ts
import { ApiProperty } from '@nestjs/swagger'
import { createChannel } from '@wisemen/nestjs-async-api'

export class UserCreatedEventContent {
  @ApiProperty({ type: String, format: 'uuid' })
  uuid: UserUuid

  @ApiProperty({ type: String, format: 'email' })
  email: string

  constructor (uuid: UserUuid, email: string) {
    this.uuid = uuid
    this.email = email
  }
}

export class UserCreatedIntegrationEvent extends IntegrationEvent {
  @ApiProperty({
    enumName: 'UserCreatedEventType',
    enum: [IntegrationEventType.USER_CREATED],

>   })

Reviewer: maken we hier niet meestal een dedicated ApiProperty decorator voor?

Author: Niet als het maar 1 mogelijke waarde is.

---
name: async-api

> description: Create documentation for NATS cloud events. Use this when creating a NATS event that gets broadcasted to a place outside of the API it was created (e.g. integration events, websocket events).

Reviewer 2: ```suggestion
description: Create documentation for cloud events. Use this when adding a (NATS) integration event.
```

---

Generate AsyncAPI 3.0.0 specification documents from type-safe channel definitions and Swagger-decorated DTOs.

# Validate event structure
- Make sure the properties of the event are decorated using '@nestjs/swagger'. 

> eg.

Reviewer 2: remove

```ts

> // src/events/user-created.integration.event.ts

Reviewer 2: confusing file path, remove or change with an actual file path

import { ApiProperty } from '@nestjs/swagger'

export class UserCreatedEventContent {
  @ApiProperty({ type: String, format: 'uuid' })
  uuid: UserUuid

>   @UserTypeApiProperty()

Reviewer 2: confusing because it's missing from the import, add something else like email

  type: UserType

  constructor (uuid: UserUuid, type: UserType) {
    this.uuid = uuid
    this.type = type
  }
}

export class UserCreatedIntegrationEvent extends IntegrationEvent {
  @ApiProperty({
    enumName: 'UserCreatedEventType',
    enum: [IntegrationEventType.USER_CREATED],
  })
  declare type: IntegrationEventType.USER_CREATED

  @ApiProperty({ type: UserCreatedEventContent })
  declare data: UserCreatedEventContent

  constructor (uuid: UserUuid, type: UserType) {
    super({
      type: IntegrationEventType.USER_CREATED,
      data: new UserCreatedEventContent(uuid, type),
      version: '0.0.1'
    })
  }
}
```

# Define a channel

> - This happens in the same file as the event definition, at the bottom of the file.

Reviewer 2: ```suggestion
- Define the channel at the bottom of the same file as the integration event definition.
```

> - Make a separate constant containing the subject of the event.

Reviewer 2: ```suggestion
- Create a constant for the subject (i.e. topic) where the event will be sent
```

  - The subject may contain parameters like 'environment', 'uuid', ... These parameters must be enclosed by curly brackets (`{}`)
- Create a channel using the `createChannel` method from the '@wisemen/nestjs-async-api' package.
  - All parameters defined in the subject must be documented in the channel config param of the method.
  - Define operations for the event. These can be 'send' or 'receive'
  

> eg.

Reviewer 2: remove

```ts

> // src/events/user-created.integration.event.ts

Reviewer 2: confusing file name

import { createChannel } from '@wisemen/nestjs-async-api'

> 

Reviewer 2: missing the definition of the event, probably best to just combine the two examples into one.
